ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java

```diff
@@ -283,38 +283,114 @@ private String formatQualifiedName(QualifiedName name) {
       return new Pair<>("(!" + exprString + ")", Schema.BOOLEAN_SCHEMA);
     }
 
+    private String nullCheckPrefix(ComparisonExpression.Type type) {
+      switch(type) {
+        case IS_DISTINCT_FROM:
+          return "(((Object)%1$s) == null || ((Object)%2$s) == null) ? "
+              + "((((Object)%1$s) == null ) ^ (((Object)%2$s) == null )) : ";
+        default:
+          return "(((Object)%1$s) == null || ((Object)%2$s) == null) ? false : ";
+      }
+    }
+
+    private String visitStringComparisonExpression(ComparisonExpression.Type type) {
+      switch (type) {
+        case EQUAL:
+          return "(%1$s.equals(%2$s))";
+        case NOT_EQUAL:
+        case IS_DISTINCT_FROM:
+          return "(!%1$s.equals(%2$s))";
+        case GREATER_THAN_OR_EQUAL:
+        case GREATER_THAN:
+        case LESS_THAN_OR_EQUAL:
+        case LESS_THAN:
+          return "(%1$s.compareTo(%2$s) " + type.getValue() + " 0)";
+        default:
+          throw new KsqlException("Unexpected string comparison: " + type.getValue());
+      }
+    }
+
+    private String visitArrayComparisonExpression(ComparisonExpression.Type type) {
+      switch (type) {
+        case EQUAL:
+          return "(java.util.Arrays.equals(%1$s, %2$s))";
```

Comment:
What do we want to happen if we have one array containing a `BIGINT` and another containing an `INT`, both with the same numeric value. Should these two arrays be equal in the same way that their two elements would be considered equal with this PR?  If yes, then we can't use `Arrays.equals`. Same applies to maps too.

Reply:
I couldn't think of a good way to do a proper value based array comparison in one statement. We could use streams to zip the two array together and then reduce, but that gets tricky when you have nested arrays. We could also pass in a comparator object that does the comparison, but that will be painful to maintain as we add types. Other systems (e.g. postgres) don't let you compare arrays with different value types. So this limitation seems acceptable. That said, @hjafarpour suggested that it would be better to implement comparisons of arrays/maps as KUDFS, which I agree with. So I'm going to disable these comparisons for now and file an issue for implementing them as KUDFs